model.py

```diff
@@ -9,43 +9,98 @@
+        self.hidden_size = hidden_size
+        self.rnn1 = nn.GRU(input_size=input_size, hidden_size=hidden_size, num_layers=1)
+        self.rnn2 = nn.GRU(input_size=hidden_size+self.embedding_dim, hidden_size=hidden_size, num_layers=max(1, num_layers-1))
         self.linear = nn.Linear(in_features=hidden_size, out_features=output_size)
+        self.attention = GaussianAttention(query_size=hidden_size,use_position=use_position == 1)
+        self.textConverter = data.Data()
+
 
-    def forward(self, strokes):
+    def forward(self, strokes, texts, textsmask):
         hidden = None
-        output, hidden = self.step(strokes, hidden)
+        textEmbedding = self.embedding(texts)
+        output, hidden = self.step(strokes, textEmbedding, textsmask, hidden)
         return output
 
-    def step(self, x, hidden):
-        x, hidden = self.rnn(x, hidden)
-        output = self.linear(x)
-        return output, hidden
+    def step(self, x, textEmbedding, textsmask, hidden):
+        if hidden is None:
+            hidden1 = None
+            positions = None
+            hidden3 = None
+        else:
+            hidden1, positions, hidden3 = hidden
+        o1, hidden1 = self.rnn1(x, hidden1)
+        context, positions = self.attention(o1, textEmbedding, textsmask, positions)
+        o2, hidden3 = self.rnn2(torch.cat((o1, context), dim=-1), hidden3)
+        output = self.linear(o2)
+
+        result = output, (hidden1, positions, hidden3)
+        return result
 
     @property
     def device(self):
         return next(self.parameters()).device
 
     @argbind.bind
-    def sample(self, batch_size=1, num_steps=100, bias=0.0, stddev=0.1):
+    def sample(self, promptStr, num_steps=120, bias=0.0, stddev=0.1):
+        batch_size = 1
+        promptEmbedding = self.embedding(torch.tensor(self.textConverter.text2code(promptStr), device=self.device).unsqueeze(dim=1))
+        promptMask = torch.ones((len(promptStr), 1), device=self.device)
         sample = torch.zeros(1, batch_size, 3, device=self.device)
         hidden = None
+        stddev = torch.log(torch.tensor([stddev, stddev]).view(1, batch_size, 2).to(self.device))
         output = []
-        for i in range(num_steps):
-            x, hidden = self.step(sample, hidden)
-            # Bias is weird now because we have a fixed stddev.  Instead of using bias, we could
-            # just adjust stddev.
-            #
-            # However, in the future, log-stddev will be part of the output of the model, and so it makes sense
-            # to apply bias independently.
-            zeros = torch.zeros(batch_size, 2, device=self.device)
-            std = torch.exp(torch.log(zeros + stddev) - bias)
-            noise = torch.normal(mean=zeros, std=std)
-            coords = x[:, :, 1:3] + noise
+        for _ in range(num_steps):
+            x, hidden = self.step(sample, promptEmbedding, promptMask, hidden)
+            def sample_from_distribution(mean, std_logits):
```

Comment:
This function is defined inside the for loop.

Reply:
Moved